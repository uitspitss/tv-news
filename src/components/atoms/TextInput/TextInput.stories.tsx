import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { TextInput, TextInputProps } from './TextInput';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: TextInput,
} as Meta;

const Template: Story<TextInputProps> = (args) => <TextInput {...args} />;

export const Input = Template.bind({});
Input.args = {
  placeholder: 'placeholder',
};
