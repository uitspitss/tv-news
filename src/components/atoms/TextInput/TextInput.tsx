import React, {
  memo,
  ComponentPropsWithRef,
  forwardRef,
  ForwardRefExoticComponent,
} from 'react';
import tw, { styled } from 'twin.macro';

export type TextInputProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * label
   */
  label: string;
  /**
   * error message
   */
  error?: string;
} & ComponentPropsWithRef<'input'>;

type Props = TextInputProps;

const Component: ForwardRefExoticComponent<Props> = forwardRef((props, ref) => {
  const { className, name, type = 'text', label, error, ...rest } = props;

  return (
    <div className={className}>
      <label htmlFor={name}>{label}</label>
      <input
        name={name}
        id={name}
        type={type}
        aria-label={label}
        aria-invalid={error ? 'true' : 'false'}
        ref={ref}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...rest}
      />
      {error && (
        <span className="error" role="alert">
          {error}
        </span>
      )}
    </div>
  );
});

const StyledComponent = styled(Component)`
  & > {
    input {
      ${tw`w-full bg-gray-100 rounded border border-gray-400 focus:outline-none focus:border-indigo-500`}
    }

    .error {
      ${tw`text-red-500`}
    }
  }
`;

/**
 * TextInput
 */
export const TextInput: ForwardRefExoticComponent<TextInputProps> = memo(
  forwardRef((props, ref) => {
    return <StyledComponent {...props} ref={ref} />;
  }),
);
