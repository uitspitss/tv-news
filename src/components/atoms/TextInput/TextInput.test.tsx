import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { TextInput } from './TextInput';

describe('TextInput', () => {
  test('snapshot test', () => {
    const { container } = render(<TextInput label="textinput label" />);
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      render(<TextInput label="textinput label" />);
    });

    test('initial view', () => {
      expect(screen.getByLabelText('textinput label')).toBeInTheDocument();
      expect(screen.getByText('textinput label')).toBeInTheDocument();
    });

    test('input, then value is input value', () => {
      userEvent.type(
        screen.getByLabelText('textinput label'),
        'input something',
      );
      expect(screen.getByLabelText('textinput label')).toHaveValue(
        'input something',
      );
    });
  });
});
