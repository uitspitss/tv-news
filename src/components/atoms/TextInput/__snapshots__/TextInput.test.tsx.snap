// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput snapshot test 1`] = `
.emotion-0>input {
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
}

.emotion-0>input:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(99, 102, 241, var(--tw-border-opacity));
}

.emotion-0>.error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <label>
      textinput label
    </label>
    <input
      aria-invalid="false"
      aria-label="textinput label"
      type="text"
    />
  </div>
</div>
`;
